### PR TITLE
Sparse dvr

### PR DESCRIPTION
**Pull request by halvarsu**
Use the library [sparse](https://sparse.pydata.org/en/latest/index.html) to represent the sparse u-matrix elements. Since `sparse` does not yet have an einsum-implementation, all `einsum`-calls have to be rewritten to `tensordot` if this is going to work. Is this a feasible and/or desirable task?

### CODE REVIEW

File: quantum_systems/sinc_dvr/one_dim/sinc_dvr.py [hunk 1]
```diff
@@ -118,50 +149,106 @@ def construct_dipole_moment(self):
...
+            assert (
+                not anti_symmetrize
+            ), "antisymmetrize only valid for sparse storage of u"
+            # call static method of superclass
+            return super(ODSincDVR, ODSincDVR).transform_two_body_elements(
+                u, C, np, C_tilde
+            )
 
     def change_basis(self, *args, **kwargs):
         super().change_basis(*args, **kwargs)
```
Comment (Schoyen) [inline quantum_systems/sinc_dvr/one_dim/sinc_dvr.py:254]: This method looks redundant :P

**Review by Schoyen — CHANGES_REQUESTED** [review]
This looks good all in all. There's only a few small comments, and maybe we need to discuss the handling of the spin-squared operator for DVR?

**Comment by Schoyen** [comment]
Hmm, come to think of it, you need to merge with master first. This pull request looks to be a little behind.

File: quantum_systems/quantum_dots/one_dim/one_dim_potentials.py [hunk 2]
```diff
@@ -43,6 +49,9 @@ def __call__(self, x):
             * (x - 0.5 * self.a) ** 2
         )
 
+    def derivative(self, x):
+        raise NotImplementedError()
+
```
[the remark this answers is not in the document]
Comment (halvarsu) [reply quantum_systems/quantum_dots/one_dim/one_dim_potentials.py:54]: Ah yes of course. This was done in the Northug-part of my master:-)

File: quantum_systems/sinc_dvr/one_dim/sinc_dvr.py [hunk 3]
```diff
@@ -5,11 +5,6 @@
 import scipy.special as spec
 
 from ... import BasisSet
```
[the remark this answers is not in the document]
Comment (halvarsu) [reply quantum_systems/sinc_dvr/one_dim/sinc_dvr.py:7]: yup. You are right, that is better with regards to pep 8 as well

File: quantum_systems/sinc_dvr/one_dim/sinc_dvr.py [hunk 4]
```diff
@@ -118,50 +149,106 @@ def construct_dipole_moment(self):
...
-        return np.eye(self.l_dvr)
-
-    def add_spin_two_body(self, _u, np):
-        if self._sparse_u:
-            return add_spin_two_body(_u, np)
+        return np.eye(self.l)
+    
+    def change_to_general_orbital_basis(self, anti_symmetrize=True):
+        if anti_symmetrize and self.u_repr == "2d":
+            self.set_u_repr("4d")
```
[the remark this answers is not in the document]
Comment (halvarsu) [reply quantum_systems/sinc_dvr/one_dim/sinc_dvr.py:179]: Makes sense, but we would have to set a limit for what "large" constitutes. Maybe `l > 100`, as this gives `size(u) = 1.6 GB` for complex-valued matrix elements?

**Comment by halvarsu** [comment]
Huh, doesn't github show merge-commits from master?

File: quantum_systems/sinc_dvr/one_dim/sinc_dvr.py (hunk 1 above)
Comment (Schoyen) [reply quantum_systems/sinc_dvr/one_dim/sinc_dvr.py:263]: Ah, I didn't see the line below :P

**Review by Schoyen — APPROVED** [review]

**Comment by Schoyen** [comment]
Lets do the merge! :champagne:

**Pull request merged**